### PR TITLE
Upsert / Insert with default_to_null boolean argument

### PR DESCRIPTION
**Pull request by sapphire008**
## What kind of change does this PR introduce?

Feature / Bug Fix: adding the `default_to_null` parameter for `upsert` and `insert` methods as presented in both JavaScript and Dart/Flutter version of the client.

## What is the current behavior?

When doing bulk insert/upsert of rows into the Postgresql table, the missing fields are default to Nulls instead of using the default value of the column. The parameter `default_to_null` allows the user to use default values from the column definition, instead of using Nulls. But this parameter is currently not exposed in the Python client, though they are available in both JavaScript and Dart/Flutter clients.

## What is the new behavior?

When `default_to_null = True` (default value), the current behavior withholds, where missing fields from the json payload will be inserted using Null. 

When setting `default_to_null = False` AND user is performing bulk insert / upsert (i.e. json payload is a list of dictionary), the rows will be inserted with default values from the table/column definitions, rather than using Nulls.

This is done by setting two things in the PostgREST call:

* Adding `columns="column1","column2"` in the query parameter of the POST call
* Adding `Prefer: missing=default` in the header of the POST call.

The documentation from both Dart/Flutter and Javascript has indicated that setting `default_to_null = False` only applies to bulk insertion/upsersion: https://supabase.com/docs/reference/javascript/insert. Therefore, to update a single record, the user can wrap the json/dict as a list, i.e. `[insert_dict]`.

* Linked issue: https://github.com/supabase-community/postgrest-py/issues/370
* Compare similar PR in Dart/Flutter client: https://github.com/supabase/supabase-flutter/pull/550/files
* Compare similar PR in JavaScript client: https://github.com/supabase/postgrest-js/pull/417/files

### CODE REVIEW

File: tests/_sync/test_request_builder.py [hunk 1]
```diff
@@ -80,6 +92,19 @@ def test_upsert(self, request_builder: SyncRequestBuilder):
         ]
         assert builder.http_method == "POST"
         assert builder.json == {"key1": "val1"}
+        
+    def test_bulk_upsert_with_default(self, request_builder: SyncRequestBuilder):
```
[the remark this answers is not in the document]
Comment (sapphire008) [reply tests/_sync/test_request_builder.py:109]: Test case added

File: tests/_async/test_request_builder.py [hunk 2]
```diff
@@ -81,6 +93,19 @@ def test_upsert(self, request_builder: AsyncRequestBuilder):
         assert builder.http_method == "POST"
         assert builder.json == {"key1": "val1"}
 
+    def test_bulk_upsert_with_default(self, request_builder: AsyncRequestBuilder):
```
[the remark this answers is not in the document]
Comment (sapphire008) [reply tests/_async/test_request_builder.py:109]: Test case added

File: postgrest/base_request_builder.py [hunk 3]
```diff
@@ -61,38 +66,51 @@ def pre_select(
 
 
 def pre_insert(
```
[the remark this answers is not in the document]
Comment (sapphire008) [reply postgrest/base_request_builder.py:68]: This is to keep the logical structure consistent with that of Dart and Javascript library.

File: postgrest/base_request_builder.py [hunk 4]
```diff
@@ -45,6 +45,11 @@ class QueryArgs(NamedTuple):
     headers: Headers
     json: Dict[Any, Any]
 
+def _unique_columns(json: List[Dict]):
+    unique_keys = {key for row in json for key in row.keys()}
+    columns = ",".join([f'"{k}"' for k in unique_keys])
+    return columns
```
[the remark this answers is not in the document]
Comment (sapphire008) [reply postgrest/base_request_builder.py:51]: Sanitation should be performed by `QueryParams` class from `httpx` which helps create the query parameters Implementation should be on-par with that of Javascript and Dart.

File: tests/_sync/test_request_builder.py [hunk 5]
```diff
@@ -70,7 +70,19 @@ def test_insert_with_upsert(self, request_builder: SyncRequestBuilder):
         ]
         assert builder.http_method == "POST"
         assert builder.json == {"key1": "val1"}
-
+        
+    def test_bullk_insert_using_default(self, request_builder: SyncRequestBuilder):
```
[the remark this answers is not in the document]
Comment (sapphire008) [reply tests/_sync/test_request_builder.py]: renamed

File: tests/_async/test_request_builder.py [hunk 6]
```diff
@@ -70,6 +70,18 @@ def test_insert_with_upsert(self, request_builder: AsyncRequestBuilder):
         ]
         assert builder.http_method == "POST"
         assert builder.json == {"key1": "val1"}
+    
+    def test_bullk_insert_using_default(self, request_builder: AsyncRequestBuilder):
```
[the remark this answers is not in the document]
Comment (sapphire008) [reply tests/_async/test_request_builder.py]: renamed

**Comment by sapphire008** [comment]
Merged the newest change from master that reverted `codecov-action` version. Should be ready to try again with CI.

**Comment by sapphire008** [comment]
Fixed linting issues using pre-commit.

**Review by silentworks — APPROVED** [review]
Thank you for this contribution.

**Pull request merged**